pym/bob/layers.py

```diff
@@ -0,0 +1,193 @@
+           and oldState is not None \
+           and oldState["digest"] == newState["digest"]:
+            log("CHECKOUT: Layer " +
+                "'{}' skipped (up to date)".format(self.getName()), SKIPPED, INFO)
+            return
+
+        if not created and oldState is not None and \
+            newState["digest"] != oldState["digest"] and \
+            self.__scm.canSwitch(Scm(oldState["prop"],
+                                 self.__recipes.getRootEnv(),
+                                 overrides=[],
+                                 recipeSet=self.__recipes)):
+                ret = await invoker.executeScmSwitch(self.__scm, oldState["prop"])
+                log("SWITCH: Layer '{}' .. ok".format(self.getName()), EXECUTED, INFO)
+
+                if ret == 0:
+                    BobState().setLayerState(layerSrcPath, newState)
+                    return
+        ret = os.path.exists(layerSrcPath)
+
+        if os.path.exists(layerSrcPath) and not created:
+            atticName = datetime.datetime.now().isoformat().translate(INVALID_CHAR_TRANS) + "_" + \
+                            self.__name
+            log("ATTIC: Layer " +
+                "'{}' (move to layers.attic/{})".format(self.getName(), atticName), WARNING, WARNING)
+            atticPath = os.path.join(self.__root, "layers.attic")
+            if not os.path.isdir(atticPath):
+                os.makedirs(atticPath)
+            atticPath = os.path.join(atticPath, atticName)
+            os.rename(layerSrcPath, atticPath)
+            BobState().delLayerState(layerSrcPath)
+
+        if not os.path.isdir(layerSrcPath):
+            os.makedirs(layerSrcPath)
+        await self.__scm.invoke(invoker)
+        log("CHECKOUT: Layer " +
+                "'{}' .. ok".format(self.getName()), EXECUTED, NORMAL)
+        BobState().setLayerState(layerSrcPath, newState)
+
+    def checkout(self, loop, verbose):
+        try:
+            j = loop.create_task(self.__checkoutTask(verbose))
+            loop.run_until_complete(j)
+        except CmdFailedError as e:
+            raise BuildError(f"Failed to checkout Layer {self.getName()}: {e.what}")
+        except InvocationError as e:
+            raise BuildError(f"Failed to checkout Layer {self.getName()}: {e.what}")
+
+    def getName(self):
+        return self.__name
+
+    def loadYaml(self, path, schema):
+        if os.path.exists(path):
+            return self.__yamlCache.loadYaml(path, schema, {}, preValidate=lambda x: None)
+        return {}
+
+    def parse(self):
+        configYaml = os.path.join(self.__layerDir, "config.yaml")
+        config = self.loadYaml(configYaml, (RecipeSet.STATIC_CONFIG_SCHEMA, b''))
+        for l in config.get('layers', []):
+            scmSpec = l.getScm()
+            if scmSpec is None: continue
+            scmSpec.update({'recipe':configYaml})
+            layerScms = Scm(scmSpec,
+                            self.__recipes.getRootEnv(),
```

Comment:
Better pass an empty environment?

Reply:
I don't think this works. We need some OS-env (e.g `SSH_AGENT_*`) for the layer checkout. Use `Env(os.evironment)` might work?

Reply:
`whitelist` (and `whitelistRemove`?) should be easy. 

But for the environment we'd need make restrictions with regards to the sting-functions use as plugins are not available during layers checkout,...
I'm also not sure if `environment` is really needed. The layer-Spec is not substituted. What could be a usecase?

Reply:
Sure, if `whitelist` is sufficient then we certainly don't add `environment` for no good reason. I was under the impression that this might be needed. Anyway, I'm fine with just `whitelist`.

We cannot add `whitelistRemove` because that can't be implemented in a top-down parsing fashion. This key exists so that higher layers (higher precedence) can remove `whitelist` entries of lower layers...

Reply:
I added the `whitlist` to `config.yaml`. Passing a empty `Env` breaks the unittest as substition of the `scm`  properties is failing...
I think it's possible to rewrite the test in order to avoid the need for passing additional variables, but maybe we can also pass the command line defines to the scm?

Reply:
I did some testing on our CI. Passing variables + substitution on the scm-properties is necessary to get this to work. Locally the layer is checked out using a `git@gitserver.foo:repo.git` url. But on the gitlab-ci we need to use `https://gitlab-ci-token:$CI_JOB_TOKEN@gitserver.foo/repo.git` instead [1]. Since there are no `scmOverrides` for the layer-scms I think the only possible solution is to use
```yaml
 url: ${GITLAB_SERVER:-git@gitserver.foo:}/repo.git
```
and use `bob layers update -DGITLAB_SERVER="https://gitlab-ci-token:$CI_JOB_TOKEN@gitserver.foo/"` on the CI.

[1] https://docs.gitlab.com/ee/ci/jobs/ci_job_token.html#use-a-job-token

Reply:
This really bothers me. Your use case basically shows that something like `scmOverrides` are necessary. Approximating this with variables may be possible but this won't be a good solution when integrating other peoples layers. If we stick just to variable substitution, it basically means that you have to patch all layers so that the URL can be overridden in a way that works for you. But inventing special `scmOverrides` for layers does not feel right either... Feels like we're running in circles... :confused:

Reply:
Maybe we have to separate the layers from the `config/default.yaml` and put them into something like `layers_default`. This special config would be parsed top-down without using additional information from the other config-files. For the CI-Usecase it would be necessary to add more layers_config's via command-line or to support `include`s as well.

Reply:
I don't think we need another type of configuration file. We just have to bite the bullet and admit that some dedicated layers SCM override is necessary. I'm not sure how to handle the details, though. In any case, how about something like the following in `config.yaml`:

```yaml
layersScmOverrides:
  - match: ...
```

which applies solely to layer SCMs? Such setting applies for the current layer and all sub-layers.

We could also have `scmOverrides` in `config.yaml`. They would apply to layer SCMs and also regular SCMs in _all_ recipes. While possible, it might make things confusing because `scmOverrides` from `default.yaml` and other configuration file (`-c ...`) won't apply for layers, even though the would have the same name. That's why I think having a separate name for layer SCM overrides might be a better idea.

In the same vein, we might better add `layersWhitelist` instead of plain `whitelist` to `config.yaml`. It will only apply to layer SCMs and would prevent any confusion about the precedence of `default.yaml` `whitelist` entries.

What we still have to add is a way how to pass additional `config.yaml` style configuration file to Bob commands. Such files could only have `layersScmOverrides` and `layersWhitelist` keys, though. Anything else would not make sense IMHO.